### PR TITLE
Prometheus: Use frontend package in Prometheus DS with a feature toggle

### PR DESCRIPTION
**Pull request by bohandley**
**What is this?**

This uses `@grafana/prometheus` frontend package in core Prometheus with the feature toggle `usePrometheusFrontendPackage`.

Components that directly use the package in `module.ts`:
- DataSource
- PromQueryEditorByApp
- PromCheatSheet

Component that indirectly uses the package:

- ConfigEditor: This component still requires auth from Azure and SigV4. These will be removed at a future date (to be determined with a migration plan to external Prometheus data sources). In the meantime, we use a Config Editor component that supports these auths and at a deeper level uses components from the `@grafana/prometheus` package.

*** also updating go.work.sum after talking with @itsmylife 

**Why do we need this?**

This is part of the decoupling and decomposing work of Grafana Prometheus. 

Special notes:

There will need to be a note in the change log when this work is GA. Notes to include
- Automatic version detection is disabled
- Default support of labels with match[] over series endpoint (users with old versions of Prometheus will have to update the version and type in the Configuration Page to get support for the series endpoint.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: e2e/various-suite/exemplars.spec.ts [hunk 1]
```diff
@@ -11,7 +11,7 @@ const addDataSource = () => {
       e2e.components.DataSource.Prometheus.configPage.exemplarsAddButton().click();
       e2e.components.DataSource.Prometheus.configPage.internalLinkSwitch().check({ force: true });
       e2e.components.DataSource.Prometheus.configPage.connectionSettings().type('http://prom-url:9090');
-      e2e.components.DataSourcePicker.inputV2().click().should('have.focus');
+      cy.get('[data-testid="data-testid Data source picker select container"]').click();
```
Comment (itsmylife) [inline e2e/various-suite/exemplars.spec.ts:14]: Is this `data-testid` correct? It looks like there is something wrong to me.

File: public/app/plugins/datasource/prometheus/module.ts [hunk 2]
```diff
@@ -1,11 +1,27 @@
...
+const usePackage = config.featureToggles.usePrometheusFrontendPackage;
+
+const PrometheusDataSourceUsed = usePackage ? PrometheusDatasource : PrometheusDatasourcePackage;
+const PromQueryEditorByAppUsed = usePackage ? PromQueryEditorByApp : PromQueryEditorByAppPackage;
+const ConfigEditorUsed = usePackage ? ConfigEditor : ConfigEditorPackage;
+const PromCheatSheetUsed = usePackage ? PromCheatSheet : PromCheatSheetPackage;
+
+// @ts-ignore
+export const plugin = new DataSourcePlugin(PrometheusDataSourceUsed)
+  // @ts-ignore
```
Comment (itsmylife) [inline public/app/plugins/datasource/prometheus/module.ts:24]: Could you please add a comment here like we are gonna remove those `@ts-ignore` lines when we'll fully on prom-package. I know you can write something better :)

File: e2e/various-suite/exemplars.spec.ts (hunk 1 above)
Comment (bohandley) [reply e2e/various-suite/exemplars.spec.ts:14]: It is correct :), all of our `data-testid` attributes need the "data-testid" prefix to work as selectors in our e2e tests.

File: public/app/plugins/datasource/prometheus/module.ts (hunk 2 above)
Comment (bohandley) [reply public/app/plugins/datasource/prometheus/module.ts:24]: Yes! I will continue debugging the reason for these type errors and add a comment

File: public/app/plugins/datasource/prometheus/configuration/DataSourceHttpSettingsOverhaulPackage.tsx [hunk 3]
```diff
@@ -0,0 +1,175 @@
+import React, { ReactElement, useState } from 'react';
+
+import { DataSourceSettings } from '@grafana/data';
+import { Auth, ConnectionSettings, convertLegacyAuthProps, AuthMethod } from '@grafana/experimental';
+import { PromOptions, docsTip, overhaulStyles } from '@grafana/prometheus';
+import { SecureSocksProxySettings, useTheme2 } from '@grafana/ui';
+// NEED TO EXPORT THIS FROM GRAFANA/UI FOR EXTERNAL DS
+import { AzureAuthSettings } from '@grafana/ui/src/components/DataSourceSettings/types';
```
[the remark this answers is not in the document]
Comment (bohandley) [reply public/app/plugins/datasource/prometheus/configuration/DataSourceHttpSettingsOverhaulPackage.tsx:8]: We can create an issue to do this, but I don't think we need to do this before we do this change. We can come back to this when we migrate all vendor authentication in the next year, so I think we have some time.

[the remark this answers is not in the document]
Comment (bohandley) [reply public/app/plugins/datasource/prometheus/configuration/DataSourceHttpSettingsOverhaulPackage.tsx:8]: Issue created! https://github.com/grafana/grafana/issues/84759

**Review by itsmylife — APPROVED** [review]

File: go.work.sum [hunk 4]
```diff
@@ -409,8 +409,8 @@ github.com/grafana/gomemcache v0.0.0-20231023152154-6947259a0586 h1:/of8Z8taCPft
 github.com/grafana/gomemcache v0.0.0-20231023152154-6947259a0586/go.mod h1:PGk3RjYHpxMM8HFPhKKo+vve3DdlPUELZLSDEFehPuU=
 github.com/grafana/grafana-plugin-sdk-go v0.212.0/go.mod h1:qsI4ktDf0lig74u8SLPJf9zRdVxWV/W4Wi+Ox6gifgs=
 github.com/grafana/grafana-plugin-sdk-go v0.215.0/go.mod h1:nBsh3jRItKQUXDF2BQkiQCPxqrsSQeb+7hiFyJTO1RE=
+github.com/grafana/grafana-plugin-sdk-go v0.216.0/go.mod h1:FdvSvOliqpVLnytM7e89zCFyYPDE6VOn9SIjVQRvVxM=
 github.com/grafana/grafana/pkg/promlib v0.0.3/go.mod h1:3El4NlsfALz8QQCbEGHGFvJUG+538QLMuALRhZ3pcoo=
-github.com/grafana/grafana/pkg/promlib v0.0.4/go.mod h1:621zjsCaFtP7mqqaiMJRN9h3SXrywAMaIPkeHvq3jsk=
```
Comment (itsmylife) [inline go.work.sum:413]: Oh no this should not be removed as the `main` is using promlib `v0.0.4` now. 
Could you please run `go mod tidy` and `go work sync` and commit the result?

**Review by itsmylife — APPROVED** [review]

File: go.work.sum (hunk 4 above)
Comment (bohandley) [reply go.work.sum:413]: Thank you for your help with this! I think @toddtreece is working on a solution for keeping go.work.sum updated. In the meantime, I appreciate the help identifying what should be in this file!

Comment (itsmylife) [reply go.work.sum:413]: This is quite confusing for me too :) I put it back here https://github.com/grafana/grafana/pull/84397/commits/26e0830bc47d25138b4b9b50f55a74f527bb1277 and the rest seems fine. Thanks for your patience and sorry for this confusion. We will have a better state hopefully soon.

**Pull request merged**